### PR TITLE
feat: Product Database

### LINKED ISSUE #52: Product Database
Establish initial product database.

- Normalize data into excel sheet with separate tabs
- Use excel features where applicable (e.g. multi-select based on data in other sheet for choices)
- Create "Product Template" with ability to specify product code for hydrating page with product data
- Create eIFU search form
- Create Product Support list
- Import product documents (non-eIFU) from wordpress 

Data Model:
- Expand data model to accommodate additional data: download links, pictures
image

Data Dump Wordpress:
[Ifus-Export-2023-April-03-1902 (1).xlsx](https://github.com/hlxsites/mammotome/files/11158230/Ifus-Export-2023-April-03-1902.1.xlsx)

### PR DESCRIPTION
**Pull request by karlpauls**
Fix #52

Test URLs:
- Before: https://main--mammotome--hlxsites.hlx.page/en/
- After: https://52-product-db--mammotome--hlxsites.hlx.page/en/ifu-index

### CODE REVIEW

File: blocks/ifus/ifus.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
+import { translate } from '../../scripts/lib-franklin.js';
+
+function createDomStructure(structure, parentElement = document.body) {
```
Comment (solaris007) [inline blocks/ifus/ifus.js:3]: we could possibly move this to `lib-franklin.js` as it is generic

File: blocks/ifus/ifus.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    if (element.position === 'prepend') {
+      parentElement.prepend(domElement);
+    } else {
+      parentElement.appendChild(domElement);
+    }
+  });
+}
+
+function populateSearch(
```
Comment (solaris007) [inline blocks/ifus/ifus.js:32]: should be `async` due to the use of `translate`

File: blocks/ifus/ifus.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  countrySelector,
+  searchButton,
+  result,
+  idProducer,
+  countryProducer,
+  assetFilter,
+  productCodes,
+  clear,
+) {
+  idProducer().forEach((value) => {
```
Comment (solaris007) [inline blocks/ifus/ifus.js:43]: option generation could be put into a generic function in order to avoid code duplication:

```
  const createOptions = (items, selector, attributes = {}) => {
    items.forEach((item) => {
      const option = document.createElement('option');
      for (const [key, value] of Object.entries(attributes(item))) {
        option.setAttribute(key, value);
      }
      option.textContent = item;
      selector.appendChild(option);
    });
  };

  createOptions(idProducer(), idSelector);
  idSelector.disabled = false;

  createOptions(
    countryProducer(),
    countrySelector,
    (country) => ({ value: country.ISO_3166_1_alpha_2_code, textContent: country['Country Name'] }),
  );
  countrySelector.disabled = false;
```

File: blocks/ifus/ifus.css [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    margin-top: 2.5rem;
+    display: block;
+    white-space: nowrap;
+    text-align: center;
+    width: 100%
+}
+
+.ifu > .ifu-selection > button {
+    background-color: #487ed9;
+    color: #fff;
```
Comment (solaris007) [inline blocks/ifus/ifus.css:74]: should probably use `--mt-background-color-primary` or similar

File: blocks/ifus/ifus.css [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    transition: all .3s;
+    margin: 0 0 15px;
+    font-family: inherit;
+    font-size: 1rem;
+    line-height: 1.5;
+}
+
+.ifu-result > div > div > div > a > button {
+    padding: 15px 50px;
+    background-color: #487ed9;
```
Comment (solaris007) [inline blocks/ifus/ifus.css:63]: should probably use `--mt-button-background-color-primary`

File: blocks/ifus/ifus.css [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    color: #fff;
```
Comment (solaris007) [inline blocks/ifus/ifus.css:64]: should probably use --mt-background-color-primary or similar

File: blocks/ifus/ifus.css [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+@media (max-width: 1024px) {
+    .ifu-result > div > div {
+        flex-direction: column;
+    }
+}
+
+.ifu > .ifu-selection > select {
+    display: block;
+    width: 100%;
+    border: 1px solid #666;
```
Comment (solaris007) [inline blocks/ifus/ifus.css:51]: should probably use local var referencing `--mt-global-color-dove-gray`

File: blocks/ifus/ifus.css [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+.ifu > div {
+    padding: 15px 20px;
+    width: 100%;
+}
+
+.ifu-result > div > div {
+    padding: 0 1rem 2rem;
+    border-bottom: 1px solid #75787b;
```
Comment (solaris007) [inline blocks/ifus/ifus.css:35]: should probably use local var referencing `--mt-global-color-accent-quaternary`

File: blocks/ifus/ifus.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+          },
+        ], result);
+      }));
+    }
+    result.classList.remove('no-result');
+  });
+}
+
+async function populate(block) {
+  const resp = await fetch('/en/dj-test.json?limit=10000');
```
Comment (solaris007) [inline blocks/ifus/ifus.js:124]: for readability and less indentation i would use a guard clause like this:

```
  const resp = await fetch('/en/dj-test.json?limit=10000');
  if (!resp.ok) {
    throw new Error(`${resp.status}: ${resp.statusText}`);
  }

  const json = await resp.json();
```

File: blocks/ifus/ifus.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  if (resp.ok) {
+    const json = await resp.json();
+    const eIFUCode = block.querySelector('#eIFU-code');
```
Comment (solaris007) [inline blocks/ifus/ifus.js:127]: you could reduce the number of vars by making it generic:

```
  const getSelectors = (prefix) => {
    return {
      code: block.querySelector(`#${prefix}-code`),
      country: block.querySelector(`#${prefix}-country`),
      search: block.querySelector(`#${prefix}-code-search`),
    };
  };

  const eIFUSelectors = getSelectors('eIFU');
  const udiSelectors = getSelectors('udi');
  const productSelectors = getSelectors('product');
```

you could then wrap `populateSearch` in a callback generator function:
```
  const createPopulateSearchCallback = (type, idProducer, assetFilter, productCodesMapper) => () => {
    populateSearch(
      type.code,
      type.country,
      type.search,
      result,
      idProducer,
      () => json.countries.data,
      assetFilter,
      productCodesMapper,
      [eIFUSelectors.code, eIFUSelectors.country, udiSelectors.code, udiSelectors.country],
    );
  };
```

later generate all your callbacks:
```
  const populateSearchCallbacks = [
    createPopulateSearchCallback(
      eIFUSelectors,
      () => json.eIFUs.data.map((value) => value.ID),
      (idSelector, countrySelector) => json.IFUAssets.data.filter(
        (asset) => asset.eIFU === idSelector.value && asset.Country === countrySelector.value,
      ),
      (idSelector) => json.eIFUToProductCodes.data
        .filter((match) => match.eIFU_ID === idSelector.value)
        .map((match) => match['Product Code']).join(', '),
    ),
    ...
]
```

and finally call the callbacks:
```
populateSearchCallbacks.forEach((callback) => callback());
```

nothing blocking, just more javascript'y - please ignore if stupid :)

File: blocks/ifus/ifus.js (hunk 1 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: Makes sense to me - I moved it to `lib-franklin.js` for now as I think we can reuse it in other places.

File: blocks/ifus/ifus.js (hunk 2 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: I don't think it needs to be `async` as the `translate` is only used in the event listener callback (which is async). However, I agree that is somewhat hard to parse - hence, I moved the callback into a separate function so that it is easier to see where the `async` part happens.

File: blocks/ifus/ifus.js (hunk 3 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: I modified your suggestion slightly but adopted the spirit. Thanks!

File: blocks/ifus/ifus.css (hunk 7 above)
Comment (karlpauls) [reply blocks/ifus/ifus.css]: Done

File: blocks/ifus/ifus.css (hunk 8 above)
Comment (karlpauls) [reply blocks/ifus/ifus.css]: Done

File: blocks/ifus/ifus.js (hunk 9 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js:110]: Done.

File: blocks/ifus/ifus.js (hunk 10 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: Yes, cleaning it up by creating a selectors object makes a lot of sense - actually, I did go a little further and added the rest of the functions as well and reuse the selectors array for the clean variable. In total, I think this makes the code a lot cleaner - thanks a lot!

File: blocks/ifus/ifus.css [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    background-color: var(--mt-button-background-color-primary);
+    color: var(--mt-background-color-primary);
+    margin-top: 2.5rem;
+    display: block;
+    white-space: nowrap;
+    text-align: center;
+    width: 100%
+}
+
+.ifu > .ifu-selection > button {
```
Comment (solaris007) [inline blocks/ifus/ifus.css:72]: this shares properties with the previous CSS block, could be optimized to:

```
.ifu-result > div > div > div > a > button,
.ifu > .ifu-selection > button {
    background-color: var(--mt-button-background-color-primary);
    color: var(--mt-background-color-primary);
    width: auto;
}

.ifu-result > div > div > div > a > button {
    padding: 15px 50px;
    margin-top: 2.5rem;
    display: block;
    white-space: nowrap;
    text-align: center;
    width: 100%;
}
```

Comment (karlpauls) [reply blocks/ifus/ifus.css:71]: Thanks, done

**Comment by karlpauls** [comment]
@solaris007, I updated the product db to reflect what we discussed - PTAL again.

File: scripts/lib-franklin.js [hunk 12]
```diff
@@ -219,6 +219,41 @@ export async function decorateIcons(element) {
   });
 }
 
+/**
```
Comment (karlpauls) [inline scripts/lib-franklin.js:222]: This is duplicated in #76 as well until we merge one of the two.

File: blocks/ifus/ifus.js [hunk 13]
```diff
@@ -0,0 +1,226 @@
+import { createDomStructure, translate } from '../../scripts/lib-franklin.js';
+
+async function handleSearch(selectors, allSelectors) {
+  selectors.result.innerHTML = '';
```
Comment (solaris007) [inline blocks/ifus/ifus.js:4]: could be changed to:
```
const { result, code, country } = selectors;
  result.innerHTML = '';
```

File: blocks/ifus/ifus.js [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+    textContent: await translate('ifuSearchTitle', 'Search results'),
+  }], selectors.result);
+
+  allSelectors.filter((entry) => entry !== selectors)
+    .flatMap((entry) => [entry.code, entry.country])
+    .forEach((value) => { value.options[0].selected = true; });
+
+  selectors.result.classList.remove('no-result');
+
+  const assets = selectors.code.value && selectors.country.value
```
Comment (solaris007) [inline blocks/ifus/ifus.js:16]: depending on previous comment, could be:
```
  const assets = code.value && country.value ? selectors.assets(code.value, country.value) : [];

```

File: blocks/ifus/ifus.js [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+  if (assets.length === 0) {
+    createDomStructure([{
+      type: 'div',
+      children: [{ type: 'strong', textContent: await translate('ifuSearchNoResult', 'No result') }],
+    }], selectors.result);
+    return;
+  }
+
+  const map = new Map();
```
Comment (solaris007) [inline blocks/ifus/ifus.js:27]: possibly rename to `assetMap`

File: blocks/ifus/ifus.js [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+          { type: 'h5', textContent: key },
+          {
+            type: 'div',
+            children: [
+              {
+                type: 'div',
+                children: [
+                  {
+                    type: 'div',
+                    textContent: selectors.productCodes(selectors.code.value),
```
Comment (solaris007) [inline blocks/ifus/ifus.js:51]: use destructured `code.value`

File: blocks/ifus/ifus.js [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+                      type: 'h6',
+                      position: 'prepend',
+                      textContent: await translate('ifuSearchProductCodes', 'Product Code(s)'),
+                    }],
+                  },
+                ],
+              },
+              {
+                type: 'div',
+                textContent: selectors.country.value,
```
Comment (solaris007) [inline blocks/ifus/ifus.js:62]: use destructured `country.value`

File: blocks/ifus/ifus.js [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                    attributes: { href: link, target: 'blank' },
+                    children: [{ type: 'button', textContent: link.substring(link.lastIndexOf('/') + 1) }],
+                  }),
+                ),
+              },
+            ],
+          },
+        ],
+      },
+    ], selectors.result);
```
Comment (solaris007) [inline blocks/ifus/ifus.js:83]: use destructured `result`

**Review by solaris007 — APPROVED** [review]
LGTM, just cosmetic readability comments

File: blocks/ifus/ifus.js (hunk 13 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: Make sense. Done

File: blocks/ifus/ifus.js (hunk 15 above)
Comment (karlpauls) [reply blocks/ifus/ifus.js]: Done

File: scripts/lib-franklin.js (hunk 12 above)
Comment (karlpauls) [reply scripts/lib-franklin.js:222]: Merging this one first.
